File: crates/ferric-rules-ffi/tests/c/multifield_copy.c
/*
 * multifield_copy.c — C-ABI regression harness for FR-CABI-008 (issue #87).
 *
 * Exercises ferric_value_multifield_copy with genuinely foreign input:
 * stack arrays, C malloc arrays, and borrowed string storage. The constructor
 * must deep-copy the complete nested tree, retain only shallow external-address
 * payload pointers, and return a tree that ferric_value_free can release using
 * Ferric's allocator provenance.
 *
 * Intended to run under AddressSanitizer + UndefinedBehaviorSanitizer via
 * `just ffi-c-harness`. Repeated late failures exercise partial-allocation
 * cleanup so LeakSanitizer, where supported (including Linux CI), detects any
 * abandoned Ferric-owned values.
 */

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "ferric.h"

static int failures = 0;

#define CHECK(cond, msg)                                              \
    do {                                                              \
        if (!(cond)) {                                                \
            fprintf(stderr, "FAIL(%s:%d): %s\n", __FILE__, __LINE__,  \
                    (msg));                                           \
            failures++;                                               \
        }                                                             \
    } while (0)

static FerricValue borrowed_text(uint32_t value_type, char *text) {
    FerricValue value = ferric_value_void();
    value.value_type = value_type;
    value.string_ptr = text;
    return value;
}

static void check_void(const FerricValue *value, const char *context) {
    CHECK(value->value_type == FERRIC_VALUE_TYPE_VOID, context);
    CHECK(value->integer == 0, context);
    CHECK(value->float_ == 0.0, context);
    CHECK(value->string_ptr == NULL, context);
    CHECK(value->multifield_ptr == NULL, context);
    CHECK(value->multifield_len == 0, context);
    CHECK(value->external_type_id == 0, context);
    CHECK(value->external_pointer == NULL, context);
}

static void test_empty_and_error_contract(void) {
    FerricValue out = ferric_value_integer(91);
    FerricError err = ferric_value_multifield_copy(NULL, 0, &out);
    CHECK(err == FERRIC_ERROR_OK, "NULL, 0 must construct an empty multifield");
    CHECK(out.value_type == FERRIC_VALUE_TYPE_MULTIFIELD,
          "empty copy must return a multifield");
    CHECK(out.multifield_ptr == NULL && out.multifield_len == 0,
          "empty copy must use a null, zero-length array");
    CHECK(ferric_value_free(&out) == FERRIC_ERROR_OK,
          "empty Ferric-owned multifield must be freeable");

    CHECK(ferric_value_multifield_copy(NULL, 0, NULL) ==
              FERRIC_ERROR_NULL_POINTER,
          "null output must return NULL_POINTER");

    out = ferric_value_integer(92);
    err = ferric_value_multifield_copy(NULL, 1, &out);
    CHECK(err == FERRIC_ERROR_NULL_POINTER,
          "NULL with non-zero length must return NULL_POINTER");
    check_void(&out, "NULL with non-zero length must leave output Void");

    FerricValue bad = ferric_value_void();
    bad.value_type = 0xDEADBEEFu;
    bad.string_ptr = (char *)&bad;
    bad.multifield_ptr = &bad;
    bad.multifield_len = SIZE_MAX;
    out = ferric_value_integer(93);
    err = ferric_value_multifield_copy(&bad, 1, &out);
    CHECK(err == FERRIC_ERROR_INVALID_ARGUMENT,
          "unknown nested tags must return INVALID_ARGUMENT");
    check_void(&out, "unknown nested tags must leave output Void");

    FerricValue null_string = ferric_value_void();
    null_string.value_type = FERRIC_VALUE_TYPE_STRING;
    out = ferric_value_integer(94);
    err = ferric_value_multifield_copy(&null_string, 1, &out);
    CHECK(err == FERRIC_ERROR_INVALID_ARGUMENT,
          "null active string pointer must return INVALID_ARGUMENT");
    check_void(&out, "null active string pointer must leave output Void");

    FerricValue null_nested = ferric_value_void();
    null_nested.value_type = FERRIC_VALUE_TYPE_MULTIFIELD;
    null_nested.multifield_len = 1;
    out = ferric_value_integer(95);
    err = ferric_value_multifield_copy(&null_nested, 1, &out);
    CHECK(err == FERRIC_ERROR_INVALID_ARGUMENT,
          "null non-empty nested array must return INVALID_ARGUMENT");
    check_void(&out, "null non-empty nested array must leave output Void");

    FerricValue cycle = ferric_value_void();
    cycle.value_type = FERRIC_VALUE_TYPE_MULTIFIELD;
    cycle.multifield_ptr = &cycle;
    cycle.multifield_len = 1;
    out = ferric_value_integer(97);
    err = ferric_value_multifield_copy(&cycle, 1, &out);
    CHECK(err == FERRIC_ERROR_INVALID_ARGUMENT,
          "cyclic input must return INVALID_ARGUMENT");
    check_void(&out, "cyclic input must leave output Void");

    FerricValue overlap[2];
    overlap[0] = ferric_value_void();
    overlap[0].value_type = FERRIC_VALUE_TYPE_MULTIFIELD;
    overlap[0].multifield_ptr = &overlap[1];
    overlap[0].multifield_len = 1;
    overlap[1] = ferric_value_integer(9);
    out = ferric_value_integer(98);
    err = ferric_value_multifield_copy(overlap, 2, &out);
    CHECK(err == FERRIC_ERROR_INVALID_ARGUMENT,
          "ancestor-overlapping input must return INVALID_ARGUMENT");
    check_void(&out, "ancestor-overlapping input must leave output Void");
}

static void test_stack_and_malloc_tree(void) {
    char top_symbol[] = "alpha";
    char top_string[] = "hello";
    char *nested_symbol = (char *)malloc(sizeof("nested"));
    int external_payload = 77;

    CHECK(nested_symbol != NULL, "nested C string allocation must succeed");
    if (nested_symbol == NULL) {
        return;
    }
    memcpy(nested_symbol, "nested", sizeof("nested"));

    FerricValue *nested =
        (FerricValue *)malloc(3 * sizeof(FerricValue));
    CHECK(nested != NULL, "nested C allocation must succeed");
    if (nested == NULL) {
        free(nested_symbol);
        return;
    }
    nested[0] = borrowed_text(FERRIC_VALUE_TYPE_SYMBOL, nested_symbol);
    nested[1] = ferric_value_integer(1234);
    nested[2] = ferric_value_float(2.5);

    FerricValue source[6];
    source[0] = ferric_value_integer(42);
    source[1] = ferric_value_float(3.25);
    source[2] = borrowed_text(FERRIC_VALUE_TYPE_SYMBOL, top_symbol);
    source[3] = borrowed_text(FERRIC_VALUE_TYPE_STRING, top_string);
    source[4] = ferric_value_void();
    source[4].value_type = FERRIC_VALUE_TYPE_MULTIFIELD;
    source[4].multifield_ptr = nested;
    source[4].multifield_len = 3;
    source[5] = ferric_value_void();
    source[5].value_type = FERRIC_VALUE_TYPE_EXTERNAL_ADDRESS;
    source[5].external_type_id = 19;
    source[5].external_pointer = &external_payload;

    FerricValue out = ferric_value_void();
    FerricError err =
        ferric_value_multifield_copy(source, 6, &out);
    CHECK(err == FERRIC_ERROR_OK,
          "mixed stack/malloc tree must copy successfully");
    if (err != FERRIC_ERROR_OK) {
        free(nested);
        free(nested_symbol);
        return;
    }

    CHECK(out.value_type == FERRIC_VALUE_TYPE_MULTIFIELD &&
              out.multifield_len == 6 && out.multifield_ptr != NULL,
          "copy must return a six-element Ferric-owned multifield");
    CHECK(out.multifield_ptr != source,
          "top-level Ferric array must not alias stack input");
    CHECK(out.multifield_ptr[4].multifield_ptr != nested,
          "nested Ferric array must not alias malloc input");
    CHECK(out.multifield_ptr[2].string_ptr != top_symbol &&
              out.multifield_ptr[3].string_ptr != top_string,
          "top-level text storage must be deep-copied");
    CHECK(out.multifield_ptr[4].multifield_ptr[0].string_ptr != nested_symbol,
          "nested text storage must be deep-copied");
    CHECK(out.multifield_ptr[5].external_pointer == &external_payload &&
              out.multifield_ptr[5].external_type_id == 19,
          "external-address payload must remain shallow and caller-owned");

    source[0].integer = -1;
    nested[1].integer = -2;
    top_symbol[0] = 'X';
    top_string[0] = 'Y';
    nested_symbol[0] = 'Z';
    free(nested);
    free(nested_symbol);

    CHECK(out.multifield_ptr[0].integer == 42,
          "top-level scalar copy must survive source mutation");
    CHECK(out.multifield_ptr[4].multifield_ptr[1].integer == 1234,
          "nested scalar copy must survive source destruction");
    CHECK(strcmp(out.multifield_ptr[2].string_ptr, "alpha") == 0,
          "symbol copy must survive source mutation");
    CHECK(strcmp(out.multifield_ptr[3].string_ptr, "hello") == 0,
          "string copy must survive source mutation");
    CHECK(strcmp(out.multifield_ptr[4].multifield_ptr[0].string_ptr,
                 "nested") == 0,
          "nested symbol copy must survive source destruction");

    CHECK(ferric_value_free(&out) == FERRIC_ERROR_OK,
          "Ferric-owned mixed tree must free recursively");
    CHECK(external_payload == 77,
          "freeing the copy must not touch caller-owned external payloads");
}

static void test_partial_failure_cleanup(void) {
    char first[] = "first";
    char second[] = "second";
    char third[] = "third";

    FerricValue nested[3];
    nested[0] = borrowed_text(FERRIC_VALUE_TYPE_SYMBOL, second);
    nested[1] = borrowed_text(FERRIC_VALUE_TYPE_STRING, third);
    nested[2] = ferric_value_void();
    nested[2].value_type = 77u;
    nested[2].string_ptr = (char *)&nested[2];

    FerricValue source[2];
    source[0] = borrowed_text(FERRIC_VALUE_TYPE_STRING, first);
    source[1] = ferric_value_void();
    source[1].value_type = FERRIC_VALUE_TYPE_MULTIFIELD;
    source[1].multifield_ptr = nested;
    source[1].multifield_len = 3;

    for (size_t i = 0; i < 512; i++) {
        FerricValue out = ferric_value_integer((int64_t)i);
        FerricError err =
            ferric_value_multifield_copy(source, 2, &out);
        CHECK(err == FERRIC_ERROR_INVALID_ARGUMENT,
              "late invalid tag must reject the complete copy");
        check_void(&out, "partial-copy failure must leave output Void");
    }
}

static void test_large_repeated_copy(void) {
    enum {
        ELEMENT_COUNT = 257,
        ROUND_COUNT = 64
    };
    char text[] = "large-borrowed-text";

    for (size_t round = 0; round < ROUND_COUNT; round++) {
        FerricValue *source =
            (FerricValue *)malloc(ELEMENT_COUNT * sizeof(FerricValue));
        CHECK(source != NULL, "large C source allocation must succeed");
        if (source == NULL) {
            return;
        }
        for (size_t i = 0; i < ELEMENT_COUNT; i++) {
            if (i % 3 == 0) {
                source[i] =
                    borrowed_text(FERRIC_VALUE_TYPE_STRING, text);
            } else {
                source[i] = ferric_value_integer((int64_t)(round + i));
            }
        }

        FerricValue out = ferric_value_void();
        FerricError err =
            ferric_value_multifield_copy(source, ELEMENT_COUNT, &out);
        CHECK(err == FERRIC_ERROR_OK,
              "large borrowed array must copy successfully");
        free(source);
        if (err != FERRIC_ERROR_OK) {
            continue;
        }

        text[0] = 'X';
        CHECK(out.multifield_len == ELEMENT_COUNT,
              "large copy must preserve element count");
        CHECK(strcmp(out.multifield_ptr[0].string_ptr,
                     "large-borrowed-text") == 0,
              "large copy must own independent string bytes");
        CHECK(out.multifield_ptr[ELEMENT_COUNT - 1].integer ==
                  (int64_t)(round + ELEMENT_COUNT - 1),
              "large copy must preserve trailing scalar");
        text[0] = 'l';

        CHECK(ferric_value_free(&out) == FERRIC_ERROR_OK,
              "large Ferric-owned tree must free recursively");
    }
}

int main(void) {
    test_empty_and_error_contract();
    test_stack_and_malloc_tree();
    test_partial_failure_cleanup();
    test_large_repeated_copy();

    if (failures == 0) {
        printf("multifield_copy: all checks passed\n");
        return 0;
    }
    fprintf(stderr, "multifield_copy: %d check(s) failed\n", failures);
    return 1;
}

File: crates/ferric-rules-ffi/src/tests/multifield_copy.rs
//! Regression tests for FR-CABI-008 (issue #87): borrowed foreign value trees
//! are deep-copied into Ferric-owned multifields with explicit provenance.

use std::ffi::CStr;
use std::ptr;

use crate::error::{ferric_last_error_global, FerricError};
use crate::types::{
    ferric_value_float, ferric_value_free, ferric_value_integer, ferric_value_multifield_copy,
    FerricValue, FerricValueType,
};

fn borrowed_text(value_type: FerricValueType, bytes: &mut [u8]) -> FerricValue {
    assert_eq!(
        bytes.last(),
        Some(&0),
        "borrowed text must be NUL-terminated"
    );
    FerricValue {
        value_type: value_type.as_raw(),
        string_ptr: bytes.as_mut_ptr().cast(),
        ..FerricValue::void()
    }
}

fn borrowed_multifield(values: &mut [FerricValue]) -> FerricValue {
    FerricValue {
        value_type: FerricValueType::Multifield.as_raw(),
        multifield_ptr: if values.is_empty() {
            ptr::null_mut()
        } else {
            values.as_mut_ptr()
        },
        multifield_len: values.len(),
        ..FerricValue::void()
    }
}

fn external_address(type_id: u32, pointer: *mut std::ffi::c_void) -> FerricValue {
    FerricValue {
        value_type: FerricValueType::ExternalAddress.as_raw(),
        external_type_id: type_id,
        external_pointer: pointer,
        ..FerricValue::void()
    }
}

fn assert_void(value: &FerricValue) {
    assert_eq!(value.value_type, FerricValueType::Void.as_raw());
    assert_eq!(value.integer, 0);
    assert_eq!(value.float.to_bits(), 0.0_f64.to_bits());
    assert!(value.string_ptr.is_null());
    assert!(value.multifield_ptr.is_null());
    assert_eq!(value.multifield_len, 0);
    assert_eq!(value.external_type_id, 0);
    assert!(value.external_pointer.is_null());
}

unsafe fn multifield_elements(value: &FerricValue) -> &[FerricValue] {
    assert_eq!(
        value.value_type,
        FerricValueType::Multifield.as_raw(),
        "expected a multifield"
    );
    if value.multifield_len == 0 {
        &[]
    } else {
        std::slice::from_raw_parts(value.multifield_ptr, value.multifield_len)
    }
}

fn nested_integer_layers(multifield_levels: usize) -> Vec<Box<[FerricValue]>> {
    let mut layers = Vec::<Box<[FerricValue]>>::with_capacity(multifield_levels + 1);
    layers.push(vec![ferric_value_integer(1)].into_boxed_slice());
    for _ in 0..multifield_levels {
        let inner = layers
            .last_mut()
            .expect("at least one input layer")
            .as_mut_ptr();
        layers.push(
            vec![FerricValue {
                value_type: FerricValueType::Multifield.as_raw(),
                multifield_ptr: inner,
                multifield_len: 1,
                ..FerricValue::void()
            }]
            .into_boxed_slice(),
        );
    }
    layers
}

#[test]
fn copy_constructs_empty_multifield_and_validates_required_pointers() {
    unsafe {
        let mut out = ferric_value_integer(91);
        assert_eq!(
            ferric_value_multifield_copy(ptr::null(), 0, &mut out),
            FerricError::Ok
        );
        assert_eq!(out.value_type, FerricValueType::Multifield.as_raw());
        assert!(out.multifield_ptr.is_null());
        assert_eq!(out.multifield_len, 0);
        assert_eq!(ferric_value_free(&mut out), FerricError::Ok);

        assert_eq!(
            ferric_value_multifield_copy(ptr::null(), 0, ptr::null_mut()),
            FerricError::NullPointer
        );

        let mut out = FerricValue {
            value_type: FerricValueType::Integer.as_raw(),
            integer: 92,
            float: 4.5,
            string_ptr: 1_usize as *mut _,
            multifield_ptr: 2_usize as *mut _,
            multifield_len: 9,
            external_type_id: 11,
            external_pointer: 3_usize as *mut _,
        };
        assert_eq!(
            ferric_value_multifield_copy(ptr::null(), 1, &mut out),
            FerricError::NullPointer
        );
        assert_void(&out);
    }
}

#[test]
fn copy_deeply_owns_mixed_nested_tree_but_not_external_payload() {
    unsafe {
        let mut symbol = b"alpha\0".to_vec();
        let mut string = b"hello\0".to_vec();
        let mut non_utf8_symbol = vec![0xff, 0xfe, 0];
        let mut external_payload = 77_i32;
        let external_payload_ptr = ptr::addr_of_mut!(external_payload).cast::<std::ffi::c_void>();

        let mut nested = [
            borrowed_text(FerricValueType::Symbol, &mut non_utf8_symbol),
            ferric_value_integer(1234),
            ferric_value_float(2.5),
        ];
        let nested_source_ptr = nested.as_mut_ptr();
        let mut source = [
            ferric_value_integer(42),
            ferric_value_float(3.25),
            borrowed_text(FerricValueType::Symbol, &mut symbol),
            borrowed_text(FerricValueType::String, &mut string),
            borrowed_multifield(&mut nested),
            external_address(19, external_payload_ptr),
            FerricValue::void(),
        ];
        let source_ptr = source.as_mut_ptr();

        let mut out = FerricValue::void();
        assert_eq!(
            ferric_value_multifield_copy(source.as_ptr(), source.len(), &mut out),
            FerricError::Ok
        );

        let copied = multifield_elements(&out);
        assert_eq!(copied.len(), source.len());
        assert_ne!(out.multifield_ptr, source_ptr);
        assert_ne!(copied[4].multifield_ptr, nested_source_ptr);
        assert_ne!(copied[2].string_ptr.cast::<u8>(), symbol.as_mut_ptr());
        assert_ne!(copied[3].string_ptr.cast::<u8>(), string.as_mut_ptr());
        assert_ne!(
            (*copied[4].multifield_ptr).string_ptr.cast::<u8>(),
            non_utf8_symbol.as_mut_ptr()
        );

        source[0].integer = -1;
        nested[1].integer = -2;
        symbol[0] = b'X';
        string[0] = b'Y';
        non_utf8_symbol[0] = 1;
        assert_eq!(source[0].integer, -1);
        assert_eq!(nested[1].integer, -2);

        let copied = multifield_elements(&out);
        assert_eq!(copied[0].integer, 42);
        assert_eq!(copied[1].float.to_bits(), 3.25_f64.to_bits());
        assert_eq!(CStr::from_ptr(copied[2].string_ptr).to_bytes(), b"alpha");
        assert_eq!(CStr::from_ptr(copied[3].string_ptr).to_bytes(), b"hello");
        let copied_nested = multifield_elements(&copied[4]);
        assert_eq!(
            copied_nested[0].value_type,
            FerricValueType::Symbol.as_raw()
        );
        assert_eq!(
            CStr::from_ptr(copied_nested[0].string_ptr).to_bytes(),
            [0xff, 0xfe]
        );
        assert_eq!(copied_nested[1].integer, 1234);
        assert_eq!(copied_nested[2].float.to_bits(), 2.5_f64.to_bits());
        assert_eq!(copied[5].external_type_id, 19);
        assert_eq!(copied[5].external_pointer, external_payload_ptr);
        assert_eq!(copied[6].value_type, FerricValueType::Void.as_raw());

        assert_eq!(ferric_value_free(&mut out), FerricError::Ok);
        assert_eq!(external_payload, 77);
    }
}

#[test]
fn copy_rejects_malformed_nested_inputs_and_leaves_void_output() {
    unsafe {
        let mut out = ferric_value_integer(1);
        let invalid = FerricValue {
            value_type: u32::MAX,
            string_ptr: 1_usize as *mut _,
            multifield_ptr: 2_usize as *mut _,
            multifield_len: usize::MAX,
            ..FerricValue::void()
        };
        assert_eq!(
            ferric_value_multifield_copy(&invalid, 1, &mut out),
            FerricError::InvalidArgument
        );
        assert_void(&out);
        let diagnostic = CStr::from_ptr(ferric_last_error_global())
            .to_string_lossy()
            .into_owned();
        assert!(
            diagnostic.contains("invalid value_type discriminant"),
            "unexpected diagnostic: {diagnostic}"
        );

        let null_string = FerricValue {
            value_type: FerricValueType::String.as_raw(),
            ..FerricValue::void()
        };
        out = ferric_value_integer(2);
        assert_eq!(
            ferric_value_multifield_copy(&null_string, 1, &mut out),
            FerricError::InvalidArgument
        );
        assert_void(&out);

        let null_nested = FerricValue {
            value_type: FerricValueType::Multifield.as_raw(),
            multifield_len: 1,
            ..FerricValue::void()
        };
        out = ferric_value_integer(3);
        assert_eq!(
            ferric_value_multifield_copy(&null_nested, 1, &mut out),
            FerricError::InvalidArgument
        );
        assert_void(&out);

        let mut cycle = FerricValue::void();
        cycle.value_type = FerricValueType::Multifield.as_raw();
        cycle.multifield_ptr = &mut cycle;
        cycle.multifield_len = 1;
        out = ferric_value_integer(4);
        assert_eq!(
            ferric_value_multifield_copy(&cycle, 1, &mut out),
            FerricError::InvalidArgument
        );
        assert_void(&out);

        let mut overlap = [FerricValue::void(), ferric_value_integer(9)];
        overlap[0].value_type = FerricValueType::Multifield.as_raw();
        overlap[0].multifield_ptr = overlap.as_mut_ptr().add(1);
        overlap[0].multifield_len = 1;
        out = ferric_value_integer(5);
        assert_eq!(
            ferric_value_multifield_copy(overlap.as_ptr(), overlap.len(), &mut out),
            FerricError::InvalidArgument
        );
        assert_void(&out);
    }
}

#[test]
fn copy_cleans_partial_nested_values_after_late_validation_failure() {
    unsafe {
        let mut first = b"first\0".to_vec();
        let mut second = b"second\0".to_vec();
        let mut third = b"third\0".to_vec();
        let mut nested = [
            borrowed_text(FerricValueType::Symbol, &mut second),
            borrowed_text(FerricValueType::String, &mut third),
            FerricValue {
                value_type: 77,
                ..FerricValue::void()
            },
        ];
        let source = [
            borrowed_text(FerricValueType::String, &mut first),
            borrowed_multifield(&mut nested),
        ];

        for index in 0..64 {
            let mut out = ferric_value_integer(index);
            assert_eq!(
                ferric_value_multifield_copy(source.as_ptr(), source.len(), &mut out),
                FerricError::InvalidArgument
            );
            assert_void(&out);
        }
    }
}

#[test]
fn copy_handles_large_borrowed_array_and_source_destruction() {
    unsafe {
        let mut text = b"borrowed-large\0".to_vec();
        let mut source = Vec::with_capacity(1025);
        for index in 0..1025 {
            if index % 5 == 0 {
                source.push(borrowed_text(FerricValueType::String, &mut text));
            } else {
                source.push(ferric_value_integer(index));
            }
        }

        let mut out = FerricValue::void();
        assert_eq!(
            ferric_value_multifield_copy(source.as_ptr(), source.len(), &mut out),
            FerricError::Ok
        );
        text[0] = b'X';
        drop(source);

        let copied = multifield_elements(&out);
        assert_eq!(copied.len(), 1025);
        assert_eq!(
            CStr::from_ptr(copied[0].string_ptr).to_bytes(),
            b"borrowed-large"
        );
        assert_eq!(copied[1024].integer, 1024);
        assert_eq!(ferric_value_free(&mut out), FerricError::Ok);
    }
}

#[test]
fn copy_rejects_excessive_acyclic_nesting_before_recursive_free_is_unsafe() {
    unsafe {
        let accepted_layers = nested_integer_layers(128);
        let accepted_outermost = accepted_layers.last().expect("outer input layer");
        let mut accepted = FerricValue::void();
        assert_eq!(
            ferric_value_multifield_copy(
                accepted_outermost.as_ptr(),
                accepted_outermost.len(),
                &mut accepted,
            ),
            FerricError::Ok
        );
        assert_eq!(ferric_value_free(&mut accepted), FerricError::Ok);

        let rejected_layers = nested_integer_layers(129);
        let rejected_outermost = rejected_layers.last().expect("outer input layer");
        let mut out = ferric_value_integer(5);
        assert_eq!(
            ferric_value_multifield_copy(
                rejected_outermost.as_ptr(),
                rejected_outermost.len(),
                &mut out,
            ),
            FerricError::InvalidArgument
        );
        assert_void(&out);
        let diagnostic = CStr::from_ptr(ferric_last_error_global()).to_string_lossy();
        assert!(
            diagnostic.contains("nesting depth exceeds maximum of 128"),
            "unexpected diagnostic: {diagnostic}"
        );
    }
}
